columnflow/production/cms/dy.py

```diff
@@ -0,0 +1,459 @@
+            not self.correction or
+            not self.unc_correction
+        ):
+            raise ValueError(
+                "Incomplete dy_weight_config: missing era, order, correction or unc_correction.",
+            )  # noqa
+
+
+@producer(
+    uses={"GenPart.*"},
+    produces={
+        "gen_dilepton_pt",
+        "gen_dilepton_vis.{pt,phi}",
+        "gen_dilepton_all.{pt,phi}",
+    },
+)
+def gen_dilepton(self, events: ak.Array, **kwargs) -> ak.Array:
+    """
+    Reconstruct the di-lepton pair from generator level info. This considers only visible final-state particles.
+    In addition it provides the four-momenta of all leptons (including neutrinos) from the hard process.
+    """
+
+    # get the absolute pdg id (to account for anti-particles) and status of the particles
+    pdg_id = abs(events.GenPart.pdgId)
+    status = events.GenPart.status
+
+    # lepton masks for DY ptll reweighting corrections
+    # -> https://indico.cern.ch/event/1495537/contributions/6359516/attachments/3014424/5315938/HLepRare_25.02.14.pdf
+    ele_mu_mask = (
+        ((pdg_id == 11) | (pdg_id == 13)) &
+        (status == 1) &
+        events.GenPart.hasFlags("fromHardProcess")
+    )
+    # taus need to have status == 2,
+    tau_mask = (
+        (pdg_id == 15) & (status == 2) & events.GenPart.hasFlags("fromHardProcess")
+    )
+
+    # lepton masks for recoil corrections
+    # -> https://indico.cern.ch/event/1495537/contributions/6359516/attachments/3014424/5315938/HLepRare_25.02.14.pdf
+    lepton_all_mask = (
+        # e, mu, taus, neutrinos
+        (
+            (pdg_id >= 11) &
+            (pdg_id <= 16) &
+            (status == 1) &
+            events.GenPart.hasFlags("fromHardProcess")
+        ) |
+        # tau decay products
+        events.GenPart.hasFlags("isDirectHardProcessTauDecayProduct")
+    )
+    lepton_vis_mask = lepton_all_mask & (
+        # no e neutrinos, mu neutrinos, or taus neutrinos
+        (pdg_id != 12) & (pdg_id != 14) | (pdg_id != 16)
+    )
+
+    # combine the masks
+    lepton_mask = ele_mu_mask | tau_mask
+    lepton_pairs = events.GenPart[lepton_mask]
+    lepton_pairs_vis = events.GenPart[lepton_vis_mask]
+    lepton_pairs_all = events.GenPart[lepton_all_mask]
+
+    # some up the four momenta of the leptons
+    lepton_pair_momenta = lepton_pairs.sum(axis=-1)
+    lepton_pair_momenta_vis = lepton_pairs_vis.sum(axis=-1)
+    lepton_pair_momenta_all = lepton_pairs_all.sum(axis=-1)
+
+    # finally, save the pt and phi of the lepton pair on generator level
+    events = set_ak_column(events, "gen_dilepton_pt", lepton_pair_momenta.pt)
+    events = set_ak_column(events, "gen_dilepton_vis.pt", lepton_pair_momenta_vis.pt)
+    events = set_ak_column(events, "gen_dilepton_vis.phi", lepton_pair_momenta_vis.phi)
+    events = set_ak_column(events, "gen_dilepton_all.pt", lepton_pair_momenta_all.pt)
+    events = set_ak_column(events, "gen_dilepton_all.phi", lepton_pair_momenta_all.phi)
```

Comment:
Since this already stores pt and phi, would it make sense to also add eta and mass here? This would make this producer a bit more general and the objects could be casted to actual LorentzVectorArrays when read again.

Reply:
Thanks, good suggestion. I changed also the `produces`.